// @sudoku/stores/globalstrategy.js
class GlobalStrategy {
    constructor() {
        this.globalStrategy = new Map();
    }
    //拼接出键
    mergeMatrixAndStrings(userGrid, strategyList) {
        strategyList.sort();
        const matrixString = userGrid.map(row => row.join("")).join("");
        const stringListString = strategyList.join("");
        return matrixString + stringListString;
    }
    //增
    addUserGrid(userGrid,strategyList, strategy) {
        let key = this.mergeMatrixAndStrings(userGrid,strategyList);
        if (!this.globalStrategy.has(key)) {
            this.globalStrategy.set(key, strategy);
        }
    }
    //get
    getStrategy(userGrid,strategyList) {
        let key = this.mergeMatrixAndStrings(userGrid,strategyList);
        return this.globalStrategy.get(key);
    }
    //查字典中是否有这个键
    hasUserGrid(userGrid,strategyList) {
        let key = this.mergeMatrixAndStrings(userGrid,strategyList);
        return this.globalStrategy.has(key) ? 1 : 0;
    }

}
export const globalStrategy = new GlobalStrategy();
