// @sudoku/stores/pastSolutionByStrategy.js
export class PastSolutionByStrategy {
    constructor() {
        // 是一个map,键为grid+strategyList,值为列表[(x, y, List:candidates[])]
        this.pastSolutionByStrategy = new Map();
    }
    //拼接出键
    mergeMatrixAndStrings(userGrid, strategyList) {
        strategyList.sort();
        const matrixString = userGrid.map(row => row.join("")).join("");
        const stringListString = strategyList.join("");
        return matrixString + stringListString;
    }
    //增
    addUserGrid(userGrid, strategyList, strategy) {
        let key = this.mergeMatrixAndStrings(userGrid,strategyList);
        if (!this.pastSolutionByStrategy.has(key)) {
            this.pastSolutionByStrategy.set(key, strategy);
        }
    }
    //get
    getSolution(userGrid, strategyList) {
        let key = this.mergeMatrixAndStrings(userGrid,strategyList);
        return this.pastSolutionByStrategy.get(key);
    }
    //查字典中是否有这个键
    hasUserGrid(userGrid, strategyList) {
        let key = this.mergeMatrixAndStrings(userGrid,strategyList);
        return this.pastSolutionByStrategy.has(key) ? 1 : 0;
    }

}
// export const pastSolutionByStrategy = new PastSolutionByStrategy();
