// @sudoku/stores/stateManager.js
import { SUDOKU_SIZE } from '@sudoku/constants';

function createStateManager() {
    const state_dict = new Map();   // key:index value:grid_state
    const index_dict = new Map();   // key:grid value:index
    // const candidate_dict = new Map();   // key:index value:candidate

    return {
        add_state(state) {
            // copy state
            let newState = [];
            for (let y = 0; y < SUDOKU_SIZE; y++) {
                newState[y] = [];
                for (let x = 0; x < SUDOKU_SIZE; x++) {
                    newState[y][x] = state[y][x];
                }
            }

            if(!index_dict.has(newState)) {    // 没有则插入
                let index = state_dict.size;
                console.log(state_dict.size);

                state_dict.set(index, newState);
                index_dict.set(state, index);
            }
            else {
                console.log('the state exist!');
            }
        }, 
        // add_candidate(candidate) {

        // },
        get_index(state) {
            return index_dict.get(state);
        },
        get_state(index) {
            return state_dict.get(index);
        },
        get_state_dict() {
            return state_dict;
        },
        clear() {
            state_dict.clear();
            index_dict.clear();
        }
    }
}

export const stateManager = createStateManager();
