// @sudoku/stores/timer.js
import { derived, writable } from 'svelte/store';

function createTimer() {
	let timerInterval = null;
	let timeBegan = null;
	let timeStopped = null;
	let stoppedDuration = 0;
	let running = false;

	const timer = writable(0);

	// The timer uses a writable store but only returns a subscribe function so it's read-only
	return {
		subscribe: timer.subscribe,

		start() {
			if (running) return;

			if (timeBegan === null) {
				this.reset();
				timeBegan = Date.now();
			}

			if (timeStopped !== null) {
				stoppedDuration += (Date.now() - timeStopped);
			}

			timerInterval = setInterval(() => {
				timer.set(Date.now() - timeBegan - stoppedDuration);
			}, 10);
			running = true;
		},

		stop() {
			running = false;
			timeStopped = Date.now();
			clearInterval(timerInterval);
		},

		reset() {
			running = false;
			clearInterval(timerInterval);
			stoppedDuration = 0;
			timeBegan = null;
			timeStopped = null;
			timer.set(0);
		},
	};
}

export const timer = createTimer();
export const timerStr = derived(timer, $timer => {
	const str = new Date($timer).toISOString().substr(11, 8);

	if (str.substr(0, 2) === '00') {
		return str.substr(3);
	}

	return str;
});
